include/vcpkg/binarycaching.h

```diff
@@ -9,53 +9,131 @@
+    private:
+        CacheStatusState m_status;
+
+        union
+        {
+            // The set of providers who know they do not have the associated cache entry.
+            // Flat vector set because N is tiny.
+            std::vector<const IBinaryProvider*> m_known_unavailable_providers; // active iff m_status == unknown
+
+            // The provider who affirmatively has the associated cache entry.
+            const IBinaryProvider* m_available_provider; // active iff m_status == available
+        };
     };
 
     struct IBinaryProvider
     {
         virtual ~IBinaryProvider() = default;
 
         /// Attempts to restore the package referenced by `action` into the packages directory.
-        virtual RestoreResult try_restore(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
-
-        /// Called upon a successful build of `action`
-        virtual void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
-
-        /// <summary>Gives the BinaryProvider an opportunity to batch any downloading or server communication for
-        /// executing `plan`.</summary>
-        /// <remarks>Must only be called once for a given binary provider instance</remarks>
-        /// <param name="actions">InOut vector of actions to be prefetched</param>
+        /// Prerequisite: action has a package_abi()
+        virtual RestoreResult try_restore(const VcpkgPaths& paths,
+                                          const Dependencies::InstallPlanAction& action) const = 0;
+
+        /// Called upon a successful build of `action` to store those contents in the binary cache.
+        /// Prerequisite: action has a package_abi()
+        virtual void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) const = 0;
+
+        /// Gives the IBinaryProvider an opportunity to batch any downloading or server communication for
+        /// executing `actions`.
+        /// `cache_status` is a vector with the same number of entries of actions, where each index corresponds
+        /// to the action at the same index in `actions`. The provider must mark the cache status as appropriate.
         virtual void prefetch(const VcpkgPaths& paths,
-                              std::vector<const Dependencies::InstallPlanAction*>& actions) = 0;
-
-        /// <summary>Requests the result of <c>try_restore()</c> without actually downloading the package. Used by CI to
-        /// determine missing packages.</summary>
-        /// <param name="results_map">InOut map to track the restored packages. Should be initialized to
-        /// <c>{&amp;action, RestoreResult::missing}</c> for all install actions</param>
-        virtual void precheck(
-            const VcpkgPaths& paths,
-            std::unordered_map<const Dependencies::InstallPlanAction*, RestoreResult>& results_map) = 0;
+                              const std::vector<Dependencies::InstallPlanAction>& actions,
```

Comment:
Because this api no longer mutates, this could be downgraded to a
```suggestion
                              View<Dependencies::InstallPlanAction> actions,
```